tests/amethyst_test/src/amethyst_application.rs

```diff
@@ -0,0 +1,1287 @@
+        resource_add_fns: Vec<FnResourceAdd>,
+    ) -> Result<Application<'static, GameData<'static, 'static>, E>>
+    where
+        SLocal: State<GameData<'static, 'static>, E> + 'static,
+    {
+        let mut application_builder =
+            Application::build(AmethystApplication::assets_dir(), first_state)?;
+        {
+            let world = &mut application_builder.world;
+            for mut function in resource_add_fns {
+                function(world);
+            }
+        }
+        application_builder.build(game_data)
+    }
+
+    /// Runs the application and returns `Ok(())` if nothing went wrong.
+    ///
+    /// This method should be called instead of the `.build()` method if the application is to be
+    /// run, as this avoids a segfault on Linux when using the GL software renderer.
+    pub fn run(self) -> Result<()> {
+        let params = (self.bundle_add_fns, self.resource_add_fns, self.state_fns);
+
+        let render = self.render;
+
+        // Run in a sub thread due to mesa's threading issues with GL software rendering
+        // See: <https://users.rust-lang.org/t/trouble-identifying-cause-of-segfault/18096>
+        thread::spawn(move || -> Result<()> {
+            amethyst::start_logger(Default::default());
+
+            if render {
+                let guard = X11_GL_MUTEX.lock().unwrap();
+
+                // Note: if this panics, the Mutex is poisoned.
+                // Unfortunately we cannot catch panics, as the application is `!UnwindSafe`
+                //
+                // We have to build the application after acquiring the lock because the window is
+                // already instantiated during the build.
+                //
+                // The mutex greatly reduces, but does not eliminate X11 window initialization
+                // errors from happening:
+                //
+                // * <https://github.com/tomaka/glutin/issues/1034> can still happen
+                // * <https://github.com/tomaka/glutin/issues/1038> may be completely removed
+                Self::build_internal(params)?.run();
+
+                drop(guard);
+            } else {
+                Self::build_internal(params)?.run();
+            }
+
+            Ok(())
+        })
+        .join()
+        .expect("Failed to run Amethyst application")
+    }
+}
+
+impl<T, E> AmethystApplication<T, E>
+where
+    T: GameUpdate,
+    E: Send + Sync + 'static,
+{
+    /// Use the specified custom event type instead of `()`.
+    ///
+    /// This **must** be invoked before any of the `.with_*()` function calls as the custom event
+    /// type parameter is changed, so we are unable to bring any of the existing parameters across.
+    ///
+    /// # Type Parameters
+    ///
+    /// * `ELocal`: Type used for custom events.
+    pub fn with_custom_event_type<ELocal>(self) -> AmethystApplication<T, ELocal>
+    where
+        ELocal: Send + Sync + 'static,
```

Comment:
I don't like these parameter names. Try to use just a single letter or something that doesn't use two consecutive upper case letters.

Reply:
https://rust-lang-nursery.github.io/api-guidelines/naming.html

Reply:
cool didn't realize they doc'ed this